strict-version-matcher-plugin/src/main/java/com/google/android/gms/dependencies/DependencyInspector.java

```diff
@@ -0,0 +1,288 @@
+      }
+    }
+  }
+
+  @NotNull
+  private String getErrorMessage(@Nonnull Dependency dep, @Nonnull ArtifactVersion resolvedVersion,
+                                 @Nonnull Collection<Node> depPaths) {
+    StringBuilder errorMessage = new StringBuilder("In project '" + projectName +
+        "' a resolved Google Play services library dependency depends on another at an exact " +
+        "version (e.g. \"" + dep.getToArtifactVersionString() + "\" " +
+        "), but isn't being resolved to that version. Behavior exhibited by the library will " +
+        "be unknown. \n\nDependency failing: " + dep.getDisplayString() + ", but " +
+        dep.getToArtifact().getArtifactId() + " version was " + resolvedVersion.getVersion() +
+        ".\n\nThe following dependencies are project dependencies that are direct or have " +
+        "transitive dependencies that lead to the artifact with the issue.");
+
+    // Append the highest level dependencies into the error message using a Set to deduplicate them.
+    // The paths are different at their leaf nodes, but that information isn't being displayed.
+    HashSet<String> directDependencyStrings = new HashSet<>();
+    StringBuilder currentString = new StringBuilder();
+    for (Node node : depPaths) {
+      String[] projectNameParts =
+          node.getDependency().getFromArtifactVersion().getArtifactId().split("-");
+      if (projectNameParts[0].equals(projectNameParts[2])) {
+        currentString.append("-- Project '")
+            .append(projectNameParts[0])
+            .append("' depends onto ");
+      } else {
+        currentString.append("-- Project '")
+            .append(projectNameParts[0])
+            .append("' depends on project '")
+            .append(projectNameParts[2])
+            .append("' which depends onto ");
+      }
+      currentString.append(node.getDependency().getToArtifact().getGroupId())
+          .append(":")
+          .append(node.getDependency().getToArtifact().getArtifactId())
+          .append("@")
+          .append(node.getDependency().getToArtifactVersionString());
+
+      directDependencyStrings.add(currentString.toString());
+      currentString.delete(0, currentString.length());
+    }
+
+    // Add dependency strings to error message.
+    for (String d : directDependencyStrings) {
+      errorMessage.append("\n").append(d);
+    }
+
+    errorMessage.append("\n\n")
+        .append("For extended debugging info execute Gradle from the command line with ")
+        .append("./gradlew --info :")
+        .append(projectName)
+        .append(":assembleDebug to see the dependency paths to the artifact. ");
+
+    if (exceptionMessageAddendum != null && !"".equals(exceptionMessageAddendum.trim())) {
+      errorMessage.append(exceptionMessageAddendum);
+    }
+
+    // Keep the error from being a single line in AndroidStudio window.
+    return errorMessage.toString().replaceAll(".{120}(?=.)", "$0\n");
```

Comment:
While the comment above explains this, the line is still very much dark magic, if there is a good URL where this came from or explains it in more detail, including it would be nice.